### ISSUE TITLE
Requirements for 1Dconvolution

### ISSUE THREAD

**Issue by nihalMM** [fchollet/keras#55]
What exactly is required for providing the 1Dconvolution and 1Dpooling layers? Aren't both special cases of 2Dconvolution and 2Dpooling?

**Comment by fchollet** [comment]
Yes, and they would be implemented as such. They would be wrappers around the 2D operations, exposing them for 1D use cases in a user-friendly way. 

The API for Convolution1D would be slightly simpler than that of Convolution2D. nb_row, nb_col would be replaced a unique parameter, like filter_length. I imagine image_shape would be superfluous.

**Comment by dbonadiman** [comment]
If i didn't got them wrong there are two types of 1D convolution possible for sequences.

One is the one used in the paper:
Natural Language Processing (almost) from Scratch by Collobert and Weston.
that is implemented as temporal convolution in Torch7

Another one is the one used in that seams slightly different:
A Convolutional Neural Network for Modelling Sentences by Nal Kalchbrenner

The two are different since in the first case we have k filters (where k is the eight of the output matrix)  of dimension NxM where n is dimension of the convolution window and m is the number of features of the input sequence. the second one has M filters of Nx1 (vectors) instead that convolve independently over the features of the input sequence.
Should both be added? or do you mind to add one of them in particular?

**Comment by fchollet** [comment]
Both can certainly be useful and would probably need to be added, in a way that makes their use cases perfectly clear. It would also be neat to have reimplementations of these works as examples...

I am wondering whether having separate classes for spatial convolution and temporal convolution, like in Torch7, would be a simpler, more user-friendly API. Any thoughts on that?

**Comment by jef5ez** [comment]
I've implemented the parts of the Kalchbrenner Modelling Sentences and had them working outside of keras. I started to add it to keras here https://github.com/fchollet/keras/blob/73b587e3c3ad218deedfdc94c19ceed3b7632925/keras/layers/convolutional.py 
Haven't had a chance to test yet, but was planning to also write up some examples.

**Comment by fchollet** [comment]
@jef5ez: sounds very cool. It would definitely be interesting to see some live examples!

**Comment by phreeza** [comment]
@dbonadiman to me it sounds like the second case is simply a special form of the first case, with M=1? 

I don't think having an API for both is necessary. The constructor for a Convolution1D layer could take 3 arguments, n_filters, filter_length and n_channels. With n_filters and filter_length being mandatory, and  n_channels as a keyword argument defaulting to 1. That would give us the second case from @dbonadiman's example by default, and the first case when explicitly required.

@fchollet what is the difference between spatial and temporal convolution in torch?

**Comment by sjhddh** [comment]
@jef5ez Thank you for your contribution to extend CNN to language models.
Could I have any examples about it ? 
i.e. apply it on IMDB data, to label positive and negative reviews.

Thx.

**Comment by jef5ez** [comment]
@sjhddh unfortunately I got busy with other things and never finished working on it. My code assumed the rows/columns of the embeddings were flipped from how keras stores them. That branch is now also several hundred commits behind master and I don't know how much else has changed about keras...
